module/modmanager/manager_test.go

```diff
@@ -286,3 +280,152 @@ func TestModManagerValidation(t *testing.T) {
 	err = mgr.Close(ctx)
 	test.That(t, err, test.ShouldBeNil)
 }
+
+func TestModuleReloading(t *testing.T) {
+	ctx := context.Background()
+
+	myHelperModel := resource.NewModel("rdk", "test", "helper")
+	rNameMyHelper := generic.Named("myhelper")
+	cfgMyHelper := resource.Config{
+		Name:  "myhelper",
+		API:   generic.API,
+		Model: myHelperModel,
+	}
+	_, err := cfgMyHelper.Validate("test", resource.APITypeComponentName)
+	test.That(t, err, test.ShouldBeNil)
+
+	// This cannot use t.TempDir() as the path it gives on MacOS exceeds module.MaxSocketAddressLength.
+	parentAddr, err := os.MkdirTemp("", "viam-test-*")
+	test.That(t, err, test.ShouldBeNil)
+	defer os.RemoveAll(parentAddr)
+	parentAddr += "/parent.sock"
+
+	// These tests neither use a resource manager nor assert anything about the
+	// existence of resources in the graph. Use a dummy RemoveOrphanedResources
+	// function so orphaned resource logic does not panic.
+	dummyRemoveOrphanedResources := func(context.Context, []resource.Name) {}
+
+	exePath := utils.ResolveFile("module/testmodule/testmodule")
+	modCfg := config.Module{
+		Name:    "test-module",
+		ExePath: exePath,
+	}
+
+	t.Run("successful restart", func(t *testing.T) {
+		logger, logs := golog.NewObservedTestLogger(t)
+
+		// Precompile module to avoid timeout issues when building takes too long.
+		test.That(t, utils.BuildInDir("module/testmodule"), test.ShouldBeNil)
+
+		mgr := NewManager(parentAddr, logger, modmanageroptions.Options{
+			UntrustedEnv:            false,
+			RemoveOrphanedResources: dummyRemoveOrphanedResources,
+		})
+		err = mgr.Add(ctx, modCfg)
+		test.That(t, err, test.ShouldBeNil)
+
+		// Add helper resource and ensure "echo" works correctly.
+		h, err := mgr.AddResource(ctx, cfgMyHelper, nil)
+		test.That(t, err, test.ShouldBeNil)
+		ok := mgr.IsModularResource(rNameMyHelper)
+		test.That(t, ok, test.ShouldBeTrue)
+
+		resp, err := h.DoCommand(ctx, map[string]interface{}{"command": "echo"})
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, resp, test.ShouldNotBeNil)
+		test.That(t, resp["command"], test.ShouldEqual, "echo")
+
+		// Run 'kill_module' command through helper resource to cause module to exit
+		// with error. Assert that after module is restarted, helper is modularly
+		// managed again and remains functional.
+		_, err = h.DoCommand(ctx, map[string]interface{}{"command": "kill_module"})
+		test.That(t, err, test.ShouldNotBeNil)
+		test.That(t, err.Error(), test.ShouldContainSubstring,
+			"error reading from server")
+
+		testutils.WaitForAssertion(t, func(tb testing.TB) {
```

Comment:
Should we file a ticket for some internal system to be able to listen to changes on the robot like this? Kind of event-ish

Reply:
Yeah that's a good idea. I think it could be a nice feature both internally for testing and externally for eventual, advanced users. Filed [PRODUCT-471](https://viam.atlassian.net/browse/PRODUCT-471) (feel free to edit or move projects).